### PR TITLE
Add `doSisuAuth` option 

### PR DESCRIPTION
**Pull request by LucienHH**
This flow allows the generation of tokens for authTitles that normally wouldn't work using the normal flow, such as XboxAppIOS, XboxGamepassIOS and MinecraftJava (MinecraftLauncher).

### CODE REVIEW

File: src/MicrosoftAuthFlow.js [hunk 1]
```diff
@@ -1,190 +1,199 @@
...
-    }
-  }
-}
-
-function sha1 (data) {
-  return crypto.createHash('sha1').update(data ?? '', 'binary').digest('hex')
-}
-
-module.exports = MicrosoftAuthFlow
+const fs = require('fs')
```
Comment (rom1504) [inline src/MicrosoftAuthFlow.js:1]: avoid changing all lines ends, so we can have a diff

Comment (rom1504) [reply src/MicrosoftAuthFlow.js:1]: that will also prevent you from getting conflicts so easily

Comment (LucienHH) [reply src/MicrosoftAuthFlow.js:1]: Yeah, that's my bad on that one, here's the change in that file for easy viewing, just one added statement to the getXboxToken function.

https://github.com/PrismarineJS/prismarine-auth/pull/24/files#diff-466421b5023aca5be930859ce1246f70fb12b4abb1cfce0a168cd235ec93aa3bR120

Comment (rom1504) [reply src/MicrosoftAuthFlow.js:1]: can you add a new commit here to fix this ?

Comment (LucienHH) [reply src/MicrosoftAuthFlow.js]: Fixed

File: src/TokenManagers/XboxTokenManager.js [hunk 2]
```diff
@@ -138,6 +138,39 @@ class XboxTokenManager {
...
+    const body = JSON.stringify(payload)
+
+    const signature = this.sign(Endpoints.SisuAuthorize, '', body).toString('base64')
+
+    const headers = { Signature: signature }
+
+    const ret = await fetch(Endpoints.SisuAuthorize, { method: 'post', headers, body }).then(checkStatus)
+    debug('Sisu Auth Response', ret)
+
+    const xsts = {
```
Comment (extremeheat) [inline src/TokenManagers/XboxTokenManager.js:162]: Can this fail (that would result in missing JSON fields)? If so improved error handling here would be good, like is handled below in getXSTSTokenWithTitle.

File: src/MicrosoftAuthFlow.js [hunk 3]
```diff
@@ -118,6 +118,14 @@ class MicrosoftAuthFlow {
         const msaToken = await this.getMsaToken()
         const ut = await this.xbl.getUserToken(msaToken, !this.options.authTitle)
 
+        if (this.options.doSisuAuth) {
+          assert(this.options.authTitle !== undefined, 'Please specify an "authTitle" in Authflow constructor when using sisu authentication')
+          debug(`[xbl] Sisu flow selected, trying to authenticate with authTitle ID ${this.options.authTitle}`)
+          const deviceToken = await this.xbl.getDeviceToken(this.options)
+          const sisu = await this.xbl.doSisuAuth(msaToken, deviceToken, this.options)
+          return sisu
```
Comment (extremeheat) [inline src/MicrosoftAuthFlow.js:126]: Does this return the same payload?

File: src/common/Titles.js [hunk 4]
```diff
@@ -1,4 +1,6 @@
-module.exports = {
-  MinecraftNintendoSwitch: '00000000441cc96b',
-  MinecraftJava: '00000000402b5328'
-}
+module.exports = {
```
Comment (extremeheat) [inline src/common/Titles.js:1]: Use LF line endings

File: docs/API.md [hunk 5]
```diff
@@ -67,3 +70,10 @@ flow.getMinecraftJavaToken().then(console.log)
 * If you just want an Microsoft/Xbox token, you do not need to specify a authTitle.
 * If doing password auth, set this option to false.
 * If specifying this, you also should provide a `deviceType`, see the doc above
+
+### What does "doSisuAuth" do ?
+
+* When `doSisuAuth` is defined as true, we will generate an xsts token using the sisu flow which is utilised in some Xbox mobile apps. 
+* This flow allows generation of tokens for authTitles that normally wouldn't work using the normal flow, such as XboxAppIOS, XboxGamepassIOS and MinecraftJava. 
+* This flow will not currently work for custom Azure apps
+* When specifying this, you should also provide an `authTitle` and a corresponding `deviceType`. For example `{ authTitle: Titles.MinecraftJava, asDevice: 'Win32' }` failing this will cause a Forbidden HTTP error.
```
Comment (extremeheat) [inline docs/API.md:79]: Typo, correct field is deviceType

**Review by extremeheat — CHANGES_REQUESTED** [review]

File: src/MicrosoftAuthFlow.js (hunk 3 above)
Comment (LucienHH) [reply src/MicrosoftAuthFlow.js:125]: Yes, same as `getXSTSTokenWithTitle`

File: src/TokenManagers/XboxTokenManager.js (hunk 2 above)
Comment (LucienHH) [reply src/TokenManagers/XboxTokenManager.js:177]: Added. Sisu has the XErr code in the response headers so didn't notice it originally.

**Review by extremeheat — APPROVED** [review]

**Pull request merged**